File: src/utils/queryRouter.ts
import routesData from '../../config/query_routes.json';

export interface QueryRoute {
  id: string;
  triggers?: string[];
  query_type: string;
  must_domains?: string[];
  preferred_domains?: string[];
  preferred_sources?: string[];
  preferred_channels?: string[];
  preferred_intents?: string[];
  blocked_domains_for_top5?: string[];
  blocked_sources_for_top5?: string[];
  allow_resource_top5?: boolean;
  freshness_preference?: string;
  explanation?: string;
}

export interface RouteResult {
  raw_query: string;
  normalized_query: string;
  query_type: string;
  route_confidence: number;
  route_source: string;
  target_domains: string[];
  target_intents: string[];
  preferred_sources: string[];
  preferred_channels: string[];
  blocked_domains_for_top5: string[];
  blocked_sources_for_top5: string[];
  allow_resource_top5: boolean;
  freshness_preference: string;
}

const queryRoutes: QueryRoute[] = routesData as QueryRoute[];

export function routeQuery(rawQuery: string): RouteResult {
  const normalizedQuery = (rawQuery || '').replace(/\s+/g, ' ').trim();
  
  let bestRoute: QueryRoute = {
    id: 'general_search',
    query_type: 'general_search',
    allow_resource_top5: true,
    freshness_preference: 'none'
  };
  
  let confidence = 0.0;
  
  for (const route of queryRoutes) {
    const triggers = route.triggers || [];
    const lowerQuery = normalizedQuery.toLowerCase();
    if (triggers.some(trigger => lowerQuery.includes(trigger.toLowerCase()))) {
      bestRoute = route;
      confidence = 0.95;
      break;
    }
  }
  
  if (/^[A-Za-z]\d{6,8}$/.test(normalizedQuery)) {
    const classRoute = queryRoutes.find(r => r.id === 'class_exam_lookup');
    if (classRoute) {
      bestRoute = classRoute;
      confidence = 0.99;
    }
  }
  
  return {
    raw_query: rawQuery,
    normalized_query: normalizedQuery,
    query_type: bestRoute.query_type || 'general_search',
    route_confidence: confidence,
    route_source: 'query_routes',
    target_domains: [...(bestRoute.must_domains || []), ...(bestRoute.preferred_domains || [])],
    target_intents: bestRoute.preferred_intents || [],
    preferred_sources: bestRoute.preferred_sources || [],
    preferred_channels: bestRoute.preferred_channels || [],
    blocked_domains_for_top5: bestRoute.blocked_domains_for_top5 || [],
    blocked_sources_for_top5: bestRoute.blocked_sources_for_top5 || [],
    allow_resource_top5: bestRoute.allow_resource_top5 !== false,
    freshness_preference: bestRoute.freshness_preference || 'none'
  };
}

File: scripts/search/query_router.py
import json
import re
import os
from typing import Any, Dict, List

def load_query_routes(path: str) -> List[Dict[str, Any]]:
    if not os.path.exists(path):
        return []
    with open(path, "r", encoding="utf-8") as handle:
        return json.load(handle)

def route_query(raw_query: str, routes: List[Dict[str, Any]]) -> Dict[str, Any]:
    normalized_query = re.sub(r"\s+", " ", str(raw_query or "")).strip()
    
    # Default fallback route
    best_route = {
        "id": "general_search",
        "query_type": "general_search",
        "must_domains": [],
        "preferred_domains": [],
        "preferred_sources": [],
        "preferred_channels": [],
        "preferred_intents": [],
        "blocked_domains_for_top5": [],
        "blocked_sources_for_top5": [],
        "allow_resource_top5": True,
        "freshness_preference": "none"
    }
    
    confidence = 0.0
    
    # Find matching route
    for route in routes:
        triggers = route.get("triggers", [])
        if any(trigger.lower() in normalized_query.lower() for trigger in triggers):
            best_route = route
            confidence = 0.95
            break
            
    # Also check if class ID pattern (e.g. B250403)
    if re.search(r"^[A-Za-z]\d{6,8}$", normalized_query):
        class_route = next((r for r in routes if r.get("id") == "class_exam_lookup"), None)
        if class_route:
            best_route = class_route
            confidence = 0.99
            
    return {
        "raw_query": raw_query,
        "normalized_query": normalized_query,
        "query_type": best_route.get("query_type", "general_search"),
        "route_confidence": confidence,
        "route_source": "query_routes",
        "target_domains": best_route.get("must_domains", []) + best_route.get("preferred_domains", []),
        "target_intents": best_route.get("preferred_intents", []),
        "preferred_sources": best_route.get("preferred_sources", []),
        "preferred_channels": best_route.get("preferred_channels", []),
        "blocked_domains_for_top5": best_route.get("blocked_domains_for_top5", []),
        "blocked_sources_for_top5": best_route.get("blocked_sources_for_top5", []),
        "allow_resource_top5": best_route.get("allow_resource_top5", True),
        "freshness_preference": best_route.get("freshness_preference", "none")
    }
